test/extensions/transport_sockets/starttls/starttls_integration_test.cc

```diff
@@ -350,11 +324,7 @@ TEST_P(StartTlsIntegrationTest, SwitchToTlsFromUpstream) {
   dispatcher_->run(Event::Dispatcher::RunType::Block);
 
   // Without closing the connection, switch to tls.
-  conn_->setTransportSocket(
-      tls_context_->createTransportSocket(std::make_shared<Network::TransportSocketOptionsImpl>(
-          absl::string_view(""), std::vector<std::string>(),
-          std::vector<std::string>{"envoyalpn"})));
-  connect_callbacks_.reset();
+  ASSERT_TRUE(conn_->startSecureTransport());
```

Comment:
There are some problems with integration tests:
- Upstream StartTls is not tested as part of Envoy. It is used only as part of testing client. To test it as part of Envoy you have to create a config with upstream StartTls transport socket. This will call the factory and instantiate UpstreamStartTls. This is not done and this is why coverage tests are failing, because factory code is never executed.
- How is the UpstreamStartTls supposed to be used inside of Envoy? Which part will do a switch from cleartext to TLS? A filter? If, so you need to instantiate a filter chain and trigger an event where the filter will signal to upstream startTls to switch to TLS. See how this is done for downstream.
- Now we have two flavours of startTls. It would be nice to add tests to verify that they do not collide and can co-exists.

Reply:
Is there an easier way to accomplish this than setting up a A -> Envoy -> C integration test? As far as I'm aware, the best way to get the UpstreamStartTls loaded is as the transport socket for a cluster in the Envoy part of that chain.

Reply:
@cpakulski, can I get some more guidance here? I've been spinning my wheels trying to get some type of A -> Envoy -> C integration test going but it's not working out.

Reply:
Sure. Can you describe your use case? Which component will call startSecureTransport and under which circumstances? Original downstream starttls socket was driven by requirements for protocols which negotiate tls in-band.

Reply:
In our case a network filter that implements a custom protocol, specifically the Finagle mux protocol. We have a handle on the connection so we just need to call `->startSecureTransport()`. To test the config aspect, all we need is to be able to establish a TCP connection from the cluster, watch the bytes, and call the method. To test the functionality the changes to the end-to-end test are sufficient but I hear your concerns about being self-consistent but wrong. We could just add another test that tests the original pathway and then bring in a test that is the similar style to what I posted.

Reply:
What is hanging me up right now is I get funneled into the `fake_upstream_` stuff in the integration tests. We might be able to make the test filter much more interesting and grab the cluster, make a connection, install the filter etc, but I would have thought simple network filter should be sufficient.

Reply:
I added a new test as you asked. Let me know what you think.